File: src-tauri/src/storage.rs
use crate::db::{ConnectionGroup, ConnectionRepository, DbConnection, DbPool, GroupRepository};
use crate::security::PasswordManager;
use std::path::PathBuf;
use tauri::AppHandle;

/// 获取应用数据目录（区分开发和生产环境）
fn get_data_dir(_app_handle: &AppHandle) -> PathBuf {
    // 检查是否为开发模式（debug 构建）
    #[cfg(debug_assertions)]
    {
        // 开发环境：使用项目根目录的 .dev-data 文件夹
        // 这样开发时的数据不会干扰生产数据
        // 注意：cargo 在 src-tauri 目录下运行，所以需要向上一级
        let dev_data_dir = std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .parent() // 从 src-tauri 回到项目根目录
            .unwrap_or(&PathBuf::from("."))
            .join(".dev-data");

        // 创建目录（如果不存在），使用 unwrap 因为开发环境中几乎不会失败
        let _ = std::fs::create_dir_all(&dev_data_dir);

        println!("Using development data directory: {:?}", dev_data_dir);
        dev_data_dir
    }

    // 生产环境：使用系统标准应用数据目录
    #[cfg(not(debug_assertions))]
    {
        let app_data = app_handle
            .path()
            .app_data_dir()
            .expect("Failed to get app data dir - please ensure app data directory is accessible");

        let data_dir = app_data.join("data");
        // 创建目录（如果不存在）
        let _ = std::fs::create_dir_all(&data_dir)
            .map_err(|e| eprintln!("Warning: Failed to create data directory: {}", e));

        println!("Using production data directory: {:?}", data_dir);
        data_dir
    }
}

/// 获取数据库文件路径
fn get_db_path(app_handle: &AppHandle) -> PathBuf {
    get_data_dir(app_handle).join("connections.db")
}

/// 初始化存储系统
pub async fn init_storage(app_handle: &AppHandle) -> Result<Storage, anyhow::Error> {
    let db_path = get_db_path(app_handle);

    // 确保父目录存在
    if let Some(parent) = db_path.parent() {
        if !parent.exists() {
            std::fs::create_dir_all(parent)?;
        }
    }

    let db_path_str = db_path.to_string_lossy().to_string();

    println!("Initializing database at: {}", db_path_str);

    // 创建数据库连接池
    let pool = DbPool::new(&db_path_str).await?;

    // 运行迁移
    crate::db::run_migrations(pool.inner()).await?;

    // 创建仓库
    let connection_repo = ConnectionRepository::new(pool.clone());
    let group_repo = GroupRepository::new(pool.clone());

    Ok(Storage {
        connection_repo,
        group_repo,
        _pool: pool,
    })
}

/// 统一存储服务
#[derive(Clone)]
pub struct Storage {
    connection_repo: ConnectionRepository,
    group_repo: GroupRepository,
    _pool: DbPool,
}

impl Storage {
    /// 获取所有连接（不包含密码）
    pub async fn get_connections(&self) -> Result<Vec<DbConnection>, anyhow::Error> {
        Ok(self.connection_repo.get_all().await?)
    }

    /// 获取连接详情（包含密码）
    pub async fn get_connection_with_password(
        &self,
        id: &str,
    ) -> Result<Option<(DbConnection, Option<String>)>, anyhow::Error> {
        let conn = self.connection_repo.get_by_id(id).await?;
        if let Some(connection) = conn {
            let password = PasswordManager::get_password(self._pool.inner(), id).await?;
            Ok(Some((connection, password)))
        } else {
            Ok(None)
        }
    }

    /// 保存连接
    pub async fn save_connection(
        &self,
        conn: DbConnection,
        password: Option<&str>,
    ) -> Result<(), anyhow::Error> {
        // 保存连接配置
        self.connection_repo.save(&conn).await?;

        // 保存密码到数据库
        if let Some(pwd) = password {
            PasswordManager::save_password(self._pool.inner(), &conn.id, pwd).await?;
        }

        Ok(())
    }

    /// 更新连接
    pub async fn update_connection(
        &self,
        id: &str,
        updated: DbConnection,
        new_password: Option<&str>,
    ) -> Result<(), anyhow::Error> {
        // 如果提供了新密码则更新，否则保留原密码
        if let Some(pwd) = new_password {
            PasswordManager::save_password(self._pool.inner(), id, pwd).await?;
        }

        // 更新连接配置
        self.connection_repo.save(&updated).await?;

        Ok(())
    }

    /// 删除连接
    pub async fn delete_connection(&self, id: &str) -> Result<(), anyhow::Error> {
        // 删除密码
        PasswordManager::delete_password(self._pool.inner(), id).await?;

        // 删除连接配置
        self.connection_repo.delete(id).await?;

        Ok(())
    }

    /// 获取所有分组
    pub async fn get_groups(&self) -> Result<Vec<ConnectionGroup>, anyhow::Error> {
        Ok(self.group_repo.get_all().await?)
    }

    /// 保存分组
    pub async fn save_group(&self, group: ConnectionGroup) -> Result<(), anyhow::Error> {
        Ok(self.group_repo.save(&group).await?)
    }

    /// 删除分组
    pub async fn delete_group(&self, id: &str) -> Result<(), anyhow::Error> {
        Ok(self.group_repo.delete(id).await?)
    }

    /// 记录连接历史
    pub async fn _log_connection_history(
        &self,
        connection_id: &str,
        action: &str,
        success: bool,
        error_message: Option<&str>,
    ) -> Result<(), anyhow::Error> {
        Ok(self
            .connection_repo
            ._log_history(connection_id, action, success, error_message)
            .await?)
    }
}

File: backend/storage.go
package backend

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"idblink/backend/localdb"
)

// Storage 统一存储服务
type Storage struct {
	pool           *localdb.Pool
	connectionRepo *localdb.ConnectionRepository
	groupRepo      *localdb.GroupRepository
	snippetRepo    *localdb.SnippetRepository
	historyRepo    *localdb.HistoryRepository
	favoriteRepo   *localdb.FavoriteRepository
	aiConfigRepo   *localdb.AIConfigRepository
}

// NewStorage 创建存储服务
func NewStorage(dataDir string) (*Storage, error) {
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create data dir: %w", err)
	}

	dbPath := filepath.Join(dataDir, "connections.db")
	pool, err := localdb.NewPool(dbPath)
	if err != nil {
		return nil, fmt.Errorf("failed to create pool: %w", err)
	}

	if err := localdb.RunMigrations(pool.DB()); err != nil {
		pool.Close()
		return nil, fmt.Errorf("failed to run migrations: %w", err)
	}

	return &Storage{
		pool:           pool,
		connectionRepo: localdb.NewConnectionRepository(pool.DB()),
		groupRepo:      localdb.NewGroupRepository(pool.DB()),
		snippetRepo:    localdb.NewSnippetRepository(pool.DB()),
		historyRepo:    localdb.NewHistoryRepository(pool.DB()),
		favoriteRepo:   localdb.NewFavoriteRepository(pool.DB()),
		aiConfigRepo:   localdb.NewAIConfigRepository(pool.DB()),
	}, nil
}

// Close 关闭存储
func (s *Storage) Close() error {
	return s.pool.Close()
}

// GetConnections 获取所有连接（不包含密码）
func (s *Storage) GetConnections() ([]*localdb.DbConnection, error) {
	return s.connectionRepo.GetAll()
}

// GetConnectionWithPassword 获取连接详情（包含密码）
func (s *Storage) GetConnectionWithPassword(id string) (*localdb.DbConnection, *string, error) {
	conn, err := s.connectionRepo.GetByID(id)
	if err != nil {
		return nil, nil, err
	}
	if conn == nil {
		return nil, nil, nil
	}

	encrypted, err := s.connectionRepo.GetPassword(id)
	if err != nil {
		return nil, nil, err
	}

	var password *string
	if encrypted != "" {
		decrypted, err := DecryptPassword(encrypted)
		if err == nil {
			password = &decrypted
		}
	}

	return conn, password, nil
}

// GetSSHCredentials 获取连接的 SSH 密码和口令（解密后返回明文）
// 任一未设置则返回空字符串。
func (s *Storage) GetSSHCredentials(id string) (sshPassword, sshPassphrase string, err error) {
	encPass, encPhrase, err := s.connectionRepo.GetSSHCredentials(id)
	if err != nil {
		return "", "", err
	}
	if encPass != "" {
		if dec, e := DecryptPassword(encPass); e == nil {
			sshPassword = dec
		}
	}
	if encPhrase != "" {
		if dec, e := DecryptPassword(encPhrase); e == nil {
			sshPassphrase = dec
		}
	}
	return sshPassword, sshPassphrase, nil
}

// SaveSSHCredentials 保存连接的 SSH 密码和口令（空字符串清除已有值）
func (s *Storage) SaveSSHCredentials(id, sshPassword, sshPassphrase string) error {
	encPass, encPhrase := "", ""
	if sshPassword != "" {
		enc, err := EncryptPassword(sshPassword)
		if err != nil {
			return fmt.Errorf("encryption error: %w", err)
		}
		encPass = enc
	}
	if sshPassphrase != "" {
		enc, err := EncryptPassword(sshPassphrase)
		if err != nil {
			return fmt.Errorf("encryption error: %w", err)
		}
		encPhrase = enc
	}
	return s.connectionRepo.SaveSSHCredentials(id, encPass, encPhrase)
}

// SaveConnection 保存连接
func (s *Storage) SaveConnection(conn *localdb.DbConnection, password *string) error {
	if err := s.connectionRepo.Save(conn); err != nil {
		return err
	}

	if password != nil && *password != "" {
		encrypted, err := EncryptPassword(*password)
		if err != nil {
			return fmt.Errorf("encryption error: %w", err)
		}
		if err := s.connectionRepo.SavePassword(conn.ID, encrypted); err != nil {
			return err
		}
	}

	return nil
}

// UpdateConnection 更新连接
func (s *Storage) UpdateConnection(id string, conn *localdb.DbConnection, newPassword *string) error {
	conn.ID = id
	if err := s.connectionRepo.Save(conn); err != nil {
		return err
	}

	if newPassword != nil && *newPassword != "" {
		encrypted, err := EncryptPassword(*newPassword)
		if err != nil {
			return fmt.Errorf("encryption error: %w", err)
		}
		if err := s.connectionRepo.SavePassword(id, encrypted); err != nil {
			return err
		}
	}

	return nil
}

// DeleteConnection 删除连接
func (s *Storage) DeleteConnection(id string) error {
	if err := s.connectionRepo.DeletePassword(id); err != nil {
		return err
	}
	return s.connectionRepo.Delete(id)
}

// UpdateSortOrders 批量更新连接排序
func (s *Storage) UpdateSortOrders(orders map[string]int) error {
	return s.connectionRepo.UpdateSortOrders(orders)
}

// GetGroups 获取所有分组
func (s *Storage) GetGroups() ([]*localdb.ConnectionGroup, error) {
	return s.groupRepo.GetAll()
}

// SaveGroup 保存分组
func (s *Storage) SaveGroup(group *localdb.ConnectionGroup) error {
	return s.groupRepo.Save(group)
}

// DeleteGroup 删除分组
func (s *Storage) DeleteGroup(id string) error {
	return s.groupRepo.Delete(id)
}

// UpdateConnectionPassword 更新连接密码
func (s *Storage) UpdateConnectionPassword(id string, password string) error {
	encrypted, err := EncryptPassword(password)
	if err != nil {
		return fmt.Errorf("encryption error: %w", err)
	}
	return s.connectionRepo.SavePassword(id, encrypted)
}

// GetSnippets 获取所有代码片段
func (s *Storage) GetSnippets() ([]*localdb.Snippet, error) {
	return s.snippetRepo.GetAll()
}

// SaveSnippet 保存代码片段
func (s *Storage) SaveSnippet(snippet *localdb.Snippet) error {
	return s.snippetRepo.Save(snippet)
}

// DeleteSnippet 删除代码片段
func (s *Storage) DeleteSnippet(id string) error {
	return s.snippetRepo.Delete(id)
}

// RecordHistory 记录操作历史
func (s *Storage) RecordHistory(connID, action string, success bool, errMsg string) error {
	return s.historyRepo.Record(connID, action, success, errMsg)
}

// GetRecentHistory 获取最近的操作历史
func (s *Storage) GetRecentHistory(limit int) ([]map[string]interface{}, error) {
	return s.historyRepo.GetRecent(limit)
}

// ClearHistory 清空操作历史
func (s *Storage) ClearHistory() error {
	return s.historyRepo.Clear()
}

// GetFavorites 获取所有收藏
func (s *Storage) GetFavorites() ([]*localdb.Favorite, error) {
	return s.favoriteRepo.GetAll()
}

// SaveFavorite 保存收藏
func (s *Storage) SaveFavorite(fav *localdb.Favorite) error {
	return s.favoriteRepo.Save(fav)
}

// DeleteFavorite 删除收藏
func (s *Storage) DeleteFavorite(id string) error {
	return s.favoriteRepo.Delete(id)
}

// ==================== AI 配置 ====================

// sensitiveAIKeys 需要加密存储的 AI 配置键
var sensitiveAIKeys = map[string]bool{
	"ai.api_key": true,
}

// GetAIConfig 获取 AI 配置值（敏感键自动解密；不存在返回空字符串）
func (s *Storage) GetAIConfig(key string) (string, error) {
	val, err := s.aiConfigRepo.Get(key)
	if err != nil {
		return "", err
	}
	if val == "" {
		return "", nil
	}
	if sensitiveAIKeys[key] {
		decrypted, err := DecryptPassword(val)
		if err != nil {
			return "", fmt.Errorf("failed to decrypt ai_config[%s]: %w", key, err)
		}
		return decrypted, nil
	}
	return val, nil
}

// SetAIConfig 设置 AI 配置值（敏感键自动加密）
func (s *Storage) SetAIConfig(key, value string) error {
	if value == "" {
		return s.aiConfigRepo.Delete(key)
	}
	if sensitiveAIKeys[key] {
		encrypted, err := EncryptPassword(value)
		if err != nil {
			return fmt.Errorf("failed to encrypt ai_config[%s]: %w", key, err)
		}
		value = encrypted
	}
	return s.aiConfigRepo.Set(key, value)
}

// GetAllAIConfig 获取所有 AI 配置（敏感键返回密文，调用方需自行判断）
func (s *Storage) GetAllAIConfig() (map[string]string, error) {
	return s.aiConfigRepo.GetAll()
}

// GetAIConfigMasked 获取所有 AI 配置，敏感键返回掩码（供前端展示）
func (s *Storage) GetAIConfigMasked() (map[string]string, error) {
	all, err := s.aiConfigRepo.GetAll()
	if err != nil {
		return nil, err
	}
	result := make(map[string]string, len(all))
	for k, v := range all {
		if sensitiveAIKeys[k] {
			result[k] = maskAPIKey(v)
		} else {
			result[k] = v
		}
	}
	return result, nil
}

// maskAPIKey 将 API Key 掩码处理。
// 输入是密文，先解密再掩码；解密失败返回固定占位符（避免泄露密文长度/内容）。
func maskAPIKey(encrypted string) string {
	plain, err := DecryptPassword(encrypted)
	if err != nil {
		// 解密失败不泄露任何信息，统一返回占位符
		return "****（无法读取）"
	}
	if len(plain) <= 8 {
		return strings.Repeat("*", len(plain))
	}
	return plain[:4] + strings.Repeat("*", len(plain)-8) + plain[len(plain)-4:]
}
